### PR TITLE
Fixed bugs #499 and #500

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
This extensive change to the code addresses a couple of bugs. #499 was the bigger one, requiring the code to move maintenance of module parameters, buffers, and submodules into managed code.

Basically, when setting a parameter property such as Linear.weight, the named_parameters collection was not updated. This meant that an optimizer would not pick up the new tensor, leading to errors.

A breaking change in this PR is that named_parameters() and parameters() both return IEnumerable instances instead of arrays. This is more flexible, and also more in line with what PyTorch does.

Further, BatchNorm{123]d did not include running_mean or running_var in the state_dict() results, nor did it allow those buffers to be accessed or set.

For #500, the problem was that eval() and train() were not properly propagated to all submodules. This manifested itself as a bug to BatchNorm1d, which does not allow batch_size=1 during training, but does during evaluation. Here, too, the solution was to do the propagation in managed code rather than in the underlying native code.

** DO NOT MERGE: once the PR is accepted, I will bump the version number.

### CODE REVIEW

File: src/TorchSharp/NN/Module.cs [hunk 1]
```diff
@@ -129,6 +139,54 @@ public Module to(DeviceType deviceType, int deviceIndex = -1)
                         THSNN_Module_to_device(handle, (int)deviceType, deviceIndex);
                         torch.CheckForErrors();
 
+                        foreach (var (_, sm) in named_children()) sm.to(deviceType, deviceIndex);
+
+                        foreach (var field in this.GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance).Where(f => !f.Name.Contains("__BackingField"))) {
```
Comment (dsyme) [inline src/TorchSharp/NN/Module.cs:144]: This is very uncomfortable, and will perhaps only work for C#.  I need to understand better why we'd be resorting to such things.

Comment (dsyme) [reply src/TorchSharp/NN/Module.cs:144]: So whatever is going on here needs to be documented properly, that is, what does it actually mean for the programming model for Modules?

Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:144]: Thanks!

Specifically, what are you uncomfortable with? The "BackingField" string comparison filter, or something else?

**Comment by dsyme** [comment]
> requiring the code to move maintenance of module parameters, buffers, and submodules into managed code.
> 
> Basically, when setting a parameter property such as Linear.weight, the named_parameters collection was not updated. This meant that an optimizer would not pick up the new tensor, leading to errors.

Are you absolutely sure this is necessary?   It is not at all ideal - and must surely represent a bug in LibTorch itself, mustn't it?  Is it really the case that C++ Modules have this problem?

> Further, BatchNorm{123]d did not include running_mean or running_var in the state_dict() results, nor did it allow those buffers to be accessed or set.

Hmmm.... Those two buffers are not parameters, my loose understanding is that they meant to be part of `state_dict`.  In DiffSharp, for example, they are buffers but not parameters https://github.com/DiffSharp/DiffSharp/blob/dev/src/DiffSharp.Core/Model.BatchNorm.fs#L38

File: src/TorchSharp/NN/Module.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:144]: The logic was introduced when I added properties as valid holders of parameters and buffers. In order to avoid double-counting, I have to filter out the backing field for such properties. Now that you bring it up, I realize that it may not work the same in F#. The F# examples run to completion, but it may be double-counting (wasting time).

**Comment by NiklasGustafsson** [comment]
> Hmmm.... Those two buffers are not parameters, my loose understanding is that they meant to be part of `state_dict`. In DiffSharp, for example, they are buffers but not parameters

That is correct. Before the PR, they did not appear in state_dict(), and now they do. As buffers.

File: src/TorchSharp/NN/Module.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:144]: I guess we could avoid this mess by only allowing buffers and parameters to be held in properties, not in fields.

**Comment by NiklasGustafsson** [comment]
> Are you absolutely sure this is necessary? It is not at all ideal - and must surely represent a bug in LibTorch itself, mustn't it? Is it really the case that C++ Modules have this problem?

I don't know what to make of it, but it certainly is what's happening. Setting the 'weight' field in native code does not affect the parameters or named_parameters collections. I also note that in PyTorch, the parameters and buffers are maintained in Python code.

File: src/TorchSharp/NN/Module.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs]: @dsyme -- I pulled the support for properties holding parameters and buffers out of this PR. We can resurface it later, with proper cross-language support.

**Comment by dsyme** [comment]
> I don't know what to make of it, but it certainly is what's happening. Setting the 'weight' field in native code does not affect the parameters or named_parameters collections. I also note that in PyTorch, the parameters and buffers are maintained in Python code.

Hmmm I see.  It's really unfortunate as I'd always hoped TorchSharp could avoid this sort of code.  I ripped out a bunch of it in the revamp, since the information seemed to be available in the C++.  We should probably dig into this a little more.

**Comment by NiklasGustafsson** [comment]
@dsyme -- we can always rip it out again later, if we discover the secret to getting parameters() to be updated. The implementation does not impact the API surface area. The breaking changes to APIs were made independently of the internals.
